velox/functions/prestosql/IPAddressFunctions.h

```diff
@@ -15,14 +15,64 @@
  */
 #pragma once
 
+#include "velox/functions/Macros.h"
+#include "velox/functions/Registerer.h"
 #include "velox/functions/prestosql/types/IPAddressType.h"
 #include "velox/functions/prestosql/types/IPPrefixType.h"
 
 namespace facebook::velox::functions {
 
+template <typename T>
+struct IPPrefixFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<IPPrefix>& result,
+      const arg_type<IPAddress>& ip,
+      const arg_type<int64_t>& prefixBits) {
+    folly::ByteArray16 addrBytes;
+    memcpy(&addrBytes, &ip, ipaddress::kIPAddressBytes);
+    std::reverse(addrBytes.begin(), addrBytes.end());
+
+    result = makeIPPrefix(folly::IPAddressV6(addrBytes), prefixBits);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<IPPrefix>& result,
+      const arg_type<Varchar>& ipString,
+      const arg_type<int64_t>& prefixBits) {
+    auto tryIp = folly::IPAddress::tryFromString(ipString);
+    if (tryIp.hasError()) {
+      VELOX_USER_FAIL("Invalid IP Invalid IP address binary:{}", ipString);
+    }
+
+    result = makeIPPrefix(
+        folly::IPAddress::createIPv6(folly::IPAddress(tryIp.value())),
+        prefixBits);
+  }
+
+ private:
+  static std::tuple<int128_t, int8_t> makeIPPrefix(
+      folly::IPAddressV6 v6Addr,
+      int64_t prefixBits) {
+    auto canonicalBytes = v6Addr.isIPv4Mapped()
+        ? v6Addr.createIPv4().mask(prefixBits).createIPv6().toByteArray()
+        : v6Addr.mask(prefixBits).toByteArray();
+
+    int128_t intAddr;
+    std::reverse(canonicalBytes.begin(), canonicalBytes.end());
```

Comment:
I see we call `std::reverse` twice.
While it is fine for the initial implementation and probably not having a major performance impact, but is there a way not to `std::reverse` at all?

Reply:
I can't think of a good way because we store the IPAddress as int128_t instead of varbinary. So we need to pay 2 reverses, one to get the ip, and then another to get back the int128_t.

Since it is only 16 bytes, it _shouldn't_ be that expensive